sycl/source/backend/opencl.cpp

```diff
@@ -0,0 +1,104 @@
+//==------- opencl.cpp - SYCL OpenCL backend -------------------------------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include <CL/sycl.hpp>
+#include <detail/platform_impl.hpp>
+#include <detail/plugin.hpp>
+#include <detail/program_impl.hpp>
+#include <detail/queue_impl.hpp>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace opencl {
+using namespace detail;
+
+// Get the OpenCL plugin.
+static const plugin &getPlugin() {
+  static const plugin *OCLPlugin = nullptr;
+  if (OCLPlugin)
+    return *OCLPlugin;
+
+  const vector_class<plugin> &Plugins = pi::initialize();
+  for (const auto &Plugin : Plugins)
+    if (Plugin.getBackend() == backend::opencl) {
+      OCLPlugin = &Plugin;
+      break;
+    }
+  if (!OCLPlugin) {
+    throw runtime_error("sycl::opencl - no OpenCL plugin",
+                        PI_INVALID_OPERATION);
+  }
+  return *OCLPlugin;
+}
+
+//----------------------------------------------------------------------------
+// Implementation of opencl::make<platform>
+__SYCL_EXPORT platform make_platform(pi_native_handle NativeHandle) {
```

Comment:
All of the `make_*` functions in this file seem to be using using only the PI API. Like with `getPlugin`, could these be generalized so we could reduce the need for replicated code for other backends?

Reply:
They are indeed only using PI API (not the native API since we don't want to link the SYCL RT with native RT), but the signature of these functions can be different between various backends, e.g. in Level-Zero we needed to pass both native platform and native device to construct a SYCL device, while OpenCL only needs native device (as platform can be queried).

Reply:
Good point. I still believe some parts of each of these functions could be reused by the other backends, but you are right it may be limiting to the point of being unwarranted.

Reply:
We can re-consider re-use when we have more interop added.